File: src/main/java/ml/windleaf/wlkitsreforged/data/DataFile.java
package ml.windleaf.wlkitsreforged.data;

import ml.windleaf.wlkitsreforged.utils.FileUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;

/**
 * A class that represents a data file
 *
 * @author WindLeaf_qwq
 */
public interface DataFile<T> {
    @Nullable Object data = null;

    /**
     * Gets the name of file without suffix
     *
     * @return name
     */
    @NotNull String getName();

    /**
     * Gets the path of data file
     *
     * @return path of file
     */
    @NotNull default String getPath() {
        return FileUtil.Companion.buildModuleDataPath(getName(), getType());
    }

    /**
     * Gets the file of data
     *
     * @return file
     */
    @NotNull default File getFile() {
        return new File(getPath());
    }

    /**
     * Gets the type of data
     *
     * @return the type of data
     * @see DataType
     */
    @NotNull DataType getType();

    /**
     * Gets the data
     *
     * @return data
     */
    @NotNull T getData();

    /**
     * Saves the data
     */
    void saveData();

    /**
     * Sets the data to target data
     *
     * @param target data
     */
    void setData(T target);

    /**
     * Loads data from file
     */
    void loadDataFromFile();
}

File: src/main/java/ml/windleaf/wlkitsreforged/internal/DataFile.kt
package ml.windleaf.wlkitsreforged.internal;

import ml.windleaf.wlkitsreforged.utils.FileUtil;
import org.jetbrains.annotations.NotNull;

import java.io.File;

/**
 * A class that represents a internal file
 *
 * @author WindLeaf_qwq
 */
interface DataFile<T> {
    /**
     * Gets the name of file without suffix
     *
     * @return name
     */
    @NotNull fun getName(): String

    /**
     * Gets the path of internal file
     *
     * @return path of file
     */
    @NotNull fun getPath() = FileUtil.buildModuleDataPath(getName(), getType())

    /**
     * Gets the file of internal
     *
     * @return file
     */
    @NotNull fun getFile() = File(getPath())

    /**
     * Gets the type of internal
     *
     * @return the type of internal
     * @see DataType
     */
    @NotNull fun getType(): DataType

    /**
     * Gets the internal
     *
     * @return internal
     */
    @NotNull fun getData(): T

    /**
     * Saves the internal
     */
    fun saveData()

    /**
     * Sets the internal to target internal
     *
     * @param target internal
     */
    fun setData(@NotNull target: T)

    /**
     * Loads internal from file
     */
    fun loadDataFromFile()
}